README.md
# Python Incolume Utils

--------

_Projeto desenvolvido e administrado incolume.com.br_

![PyPI - Python Version](https://img.shields.io/pypi/pyversions/incolumepy.makefilelicense)
[![Ruff](https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/astral-sh/ruff/main/assets/badge/v2.json)](https://github.com/astral-sh/ruff)
[![Tests CI/CD](https://github.com/development-incolume/incolume.py.changelog/actions/workflows/tests-gwa-ci-cd.yml/badge.svg)](https://github.com/development-incolume/incolume.py.changelog/actions/workflows/tests-gwa-ci-cd.yml)
![PyPI - Status](https://img.shields.io/pypi/status/incolumepy.makefilelicense)
[![GitHub Actions (Tests)](https://github.com/development-incolume/incolume.py.changelog/workflows/Tests/badge.svg)](https://github.com/development-incolume/incolume.py.changelog/)
[![codecov](https://codecov.io/gh/incolumepy/incolumepy.makefilelicense/branch/main/graph/badge.svg?token=QFULL7R8HX)](https://codecov.io/gh/incolumepy/incolumepy.makefilelicense)
![PyPI](https://img.shields.io/pypi/v/incolumepy.makefilelicense)
![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/incolumepy/incolumepy.makefilelicense?logo=tag)
![PyPI - Wheel](https://img.shields.io/pypi/wheel/incolumepy.makefilelicense)
![PyPI - Implementation](https://img.shields.io/pypi/implementation/incolumepy.makefilelicense)
![PyPI - License](https://img.shields.io/pypi/l/incolumepy.makefilelicense)
!["Code style: blue"](https://img.shields.io/badge/code%20style-blue-black)
!["Code style: ruff"](https://img.shields.io/badge/code%20style-ruff-black)
[![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=4444444)](https://pycqa.github.io/isort/)
[![Docstring style: pydocstyle](https://img.shields.io/badge/%20Docstring%20Style-PyDocStyle-%231674b1?style=flat&labelColor=444444)](http://www.pydocstyle.org/en/stable/)
[![Linter: mypy](https://img.shields.io/badge/%20Linter-Mypy-%231674b1?style=flat&labelColor=4444444)](https://mypy.readthedocs.io/en/stable/)
[![Linter: pylint](https://img.shields.io/badge/%20Linter-pylint-%231674b1?style=flat&labelColor=4444444)](https://pylint.pycqa.org/en/latest/)
[![Linter: flake8](https://img.shields.io/badge/%20Linter-flake8-%231674b1?style=flat&labelColor=4444444)](https://flake8.pycqa.org/en/latest/)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/incolumepy/incolumepy.makefilelicense)
![GitHub repo size](https://img.shields.io/github/repo-size/incolumepy/incolumepy.makefilelicense)
![GitHub issues](https://img.shields.io/github/issues/incolumepy/incolumepy.makefilelicense)
![GitHub closed issues](https://img.shields.io/github/issues-closed/incolumepy/incolumepy.makefilelicense)
![GitHub closed issues by-label](https://img.shields.io/github/issues-closed/incolumepy/incolumepy.makefilelicense/enhancement)
![GitHub issues by-label](https://img.shields.io/github/issues/incolumepy/incolume.py.changelog/bug)
![GitHub issues by-label](https://img.shields.io/github/issues/incolumepy/incolume.py.changelog/enhancement)
[![Downloads](https://pepy.tech/badge/incolume-py-changelog)](https://pepy.tech/project/incolume-py-changelog)
[![Downloads](https://pepy.tech/badge/incolume-py-changelog/month)](https://pepy.tech/project/incolume-py-changelog)
[![Downloads](https://pepy.tech/badge/incolume-py-changelog/week)](https://pepy.tech/project/incolume-py-changelog)
---

Este gera o Changelog a partir dos registros encontrados em `git tag -n`.

## Instalar o pacote

### Instalação com pip
```shell
pip install incolume.py.changelog
```
### Instalação com pipenv
```shell
pipenv install incolume.py.changelog
```

### Instalação com poetry
```shell
poetry add incolume.py.changelog
```

```shell
poetry add git+https://gitlab.com/development-incolume/incolumepy.utils.git#main
```

## Atualizar o pacote
### Atualização com pip
```shell
pip install -U incolume.py.changelog
```
### Atualização com pipenv
```shell
pipenv update incolume.py.changelog
```
### Atualização com poetry
```shell
poetry update incolume.py.changelog
```

```shell
poetry update git+https://gitlab.com/development-incolume/incolumepy.utils.git#main
```

## Gerar pacote a partir dos fontes para instalação

```shell
poetry build
```

## Documentação detalhada
Exemplos da API disponíveis em [docs/api]('docs/api/index.md')



CHANGELOG.md
# CHANGELOG


All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and [Conventional Commit](https://www.conventionalcommits.org/pt-br/v1.0.0/).

This file was automatically generated for [incolumepy.utils](https://gitlab.com/development-incolume/incolumepy.utils/-/tree/2.8.0)

---


## [Unreleased]	 &#8212; 	2023-07-22:
### Added
  - Para novos recursos.;
### Fixed
  - Para qualquer correção de bug.;
### Changed
  - Para alterações em recursos existentes.;
### Deprecated
  - Para recursos que serão removidos nas próximas versões.;
### Removed
  - Para recursos removidos nesta versão.;
### Security
  - Em caso de vulnerabilidades.;
  - Em caso de vulnerabilidades.;

## [2.8.0]	 &#8212; 	2023-07-22:
### Added
  - Adicionado Unreleased/Não publicado para o número de versão conforme indicado por https://keepachangelog.com/en/1.1.0/;
  - Adicionado uma nova seção Unreleased/Não publicado no topo baseado nas tags registradas no git;
  - Implementado a tradução para labels de mudanças ptBR -> enUS;
  - Implementado tradução automática em labels de mudança pt-BR -> en-US;
  - Implementado nova função iter_logs();
### Changed
  - Permitido o uso de labels (Added|Changed|Deprecated|Removed|Fixed|Security) em pt-BR;
  - Fatorado código para changelog_body();
### Fixed
  - Formatação visual para CHANGELOG.md retirado link quebrado para 1ª release;

## [2.7.1]	 &#8212; 	2023-07-18:
### Fixed
  - Acrescentados parametros ausentes para os comandos (gcl, gchangelog, gencl, changelog);

## [2.7.0]	 &#8212; 	2023-07-18:
### Added
  - Acrescentado CLI - Command Line Interface, para o módulo incolumepy.utils.changelog (gcl, gchangelog, gencl, changelog);
### Changed
  - Adequação da formatação visual do arquivo CHANGELOG.md com https://keepachangelog.com/pt-BR/1.0.0/;
  - Atualização da documetação inicial via README;

## [2.7.0-rc.1]	 &#8212; 	2023-07-17:
### Changed
  - Atualização da documetação inicial via README;

## [2.7.0-rc.0]	 &#8212; 	2023-07-17:
### Changed
  - Adequação da formatação visual do arquivo CHANGELOG.md com https://keepachangelog.com/pt-BR/1.0.0/;

## [2.6.0]	 &#8212; 	2023-07-11:
### Added
  - Ampliado abrangência para Python3.11;
  - Documentação compilada acrescentada ao versionamento;
  - Lint pylama;
  - Compatibilidade com Python 3.11;
  - Documentação no formato HTML;
  - Ordenação de etiquetas no CHANGELOG.md;
  - Encapsulamento Mock em testes de acesso a arquivos;
  - Módulo incolumepy.utils.changelog para tratar de recuros do changelog;
  - Aderência a convensão de commit disponível em https://www.conventionalcommits.org/pt-br/v1.0.0/;
  - Aderência a Keep a Changelog disponível em https://keepachangelog.com/pt-BR/1.0.0/;
### Changed
  - 'make setup' aperfeisoado para configuração do ambiente em apenas um comando;
  - Aplicado ISO8601 para data em registros automatizados via comando make;
  - Automação de script para gerar release e patch via comando make;
  - Estilização do CHANGELOG.md;
  - Cobertura de 100% em incolumepy.utils.__init__.py;
  - Alteração nas regras githooks para melhoria de gerenciamento do projeto e garantir o versionamento semântico disponível em https://semver.org/lang/pt-BR/;
  - Estilo visual atualizado;
  - Estilização em cores para resposta de commits;
  - Branches main/master protegido contra alterações indevidas;
  - Politica de validação com Black e isort diretamente via git hooks;
  - Mensagens geradas pela ferramenta poetry autenticadas na execução;
  - 100% de cobertura atingido;
  - Complexidade ciclomática limitada (5);
  - Aualizado pacotes de dependências do projeto;
  - Melhoria nos processos automatizados;
  - Refinamento na configuração do pylint;
  - Refinamento na configuração do pylama;
### Deprecated
  - Função "incolumepy.utils.update_changelog" tornada obsoleta em favor "incolumepy.utils.changelog.update_changelog";
### Removed
  - Lint flake8 substituído por lint pylama;
  - Compatibilidade com Python 3.6;
  - Pacote coverage;
  - Pacote jupyter;
  - Refinamento da configuração com as novas funcionalidades.;
  - Suite de testes nose desativada permanentemente;

## [2.6.0rc5]	 &#8212; 	2022-12-12:
### Changed
  - Estilização em cores para resposta de commits;
  - Branches main/master protegido contra alterações indevidas;
  - Politica de validação com Black e isort diretamente via git hooks;
  - Mensagens geradas pela ferramenta poetry autenticadas na execução;

## [2.6.0rc4]	 &#8212; 	2022-12-08:
### Changed
  - 100% de cobertura atingido;
  - Complexidade ciclomática limitada (5);
  - Aualizado pacotes de dependências do projeto;
  - Melhoria nos processos automatizados;
  - Refinamento na configuração do pylint;
  - Refinamento na configuração do pylama;
### Removed
  - Pacote coverage;
  - Pacote jupyter;

## [2.6.0rc3]	 &#8212; 	2022-12-06:
### Changed
  - Refinamento da configuração com as novas funcionalidades.;

## [2.6.0rc2]	 &#8212; 	2022-12-06:
### Added
  - Lint pylama;
  - Compatibilidade com Python 3.11;
  - Documentação no formato HTML;
### Removed
  - Lint flake8 substituído por lint pylama;
  - Compatibilidade com Python 3.6;

## [2.6.0rc1]	 &#8212; 	2022-12-05:
### Added
  - Ampliado abrangência para Python3.11;
  - Documentação compilada acrescentada ao versionamento;
  - Ordenação de etiquetas no CHANGELOG.md;
  - Encapsulamento Mock em testes de acesso a arquivos;
### Changed
  - Aperfeiçoado o camando 'make setup' para configuração do ambiente em apenas uma execução;
  - Aplicado ISO8601 para data em registros automatizados via comando make;
  - Automação de script para gerar release e patch via comando make;
### Removed
  - Suite de testes nose desativada permanentemente;

## [2.6.0-rc.0]	 &#8212; 	2022-12-03:
### Added
  - Módulo incolumepy.utils.changelog para tratar de recuros do changelog;

tests/__init__.py
"""Principal module tests."""


tests/conftest.py
"""Configurate tests."""
from pathlib import Path
from tempfile import NamedTemporaryFile

import pytest


@pytest.fixture(scope='session')
def semver_regex() -> str:
    """Fixture para regex de validação do Versionamento Semântico."""
    return r'^\d+(\.\d+){2}((-\w+\.\d+)|(\w+\d+))?$'


@pytest.fixture()
def changelog_stamps():
    """Return labels Keep a Changelog."""
    return 'Added Changed Deprecated Removed Fixed Security'.upper().strip()


@pytest.fixture()
def return_git_tag():
    """Return fake `git tag -n`."""
    return """1.0.0 Added: Fake record; other fake; Fixed: Fake fixed
    1.3.0 Fixed: Fake record; other fake; Changed: Fake fixed
    1.5.0 Added: Fake record; other fake; Fixed: Fake fixed
    2.2.0 Security: Fake record; other record; Fake fixed"""


@pytest.fixture()
def file_temp():
    """Generate aleatory filename into tempdir for tests."""
    return Path(NamedTemporaryFile(delete=False).name)


tests/test_cli.py
"""Test module for cli."""
from click.testing import CliRunner
from incolume.py.changelog import cli
import pytest
import os
from pathlib import Path


@pytest.fixture()
def cli_runner() -> CliRunner:
    """Fixture to CliRunner."""
    return CliRunner()


@pytest.mark.parametrize(
    'envvar entrance expected'.split(),
    [
        pytest.param('', {'args': 'Yoda'}, 'Oi Yoda!\n'),
        pytest.param('Yoda', {'args': ''}, 'Oi Yoda!\n'),
        pytest.param('Visitante', {'args': 'obiwan'}, 'Oi Obiwan!\n'),
    ],
)
def test_gretting(
    cli_runner: CliRunner,
    envvar: str,
    entrance: dict,
    expected: str,
) -> None:
    """Test cli gretting."""
    if envvar:
        os.environ['USERNAME'] = envvar
    result = cli_runner.invoke(cli.greeting, **entrance)
    assert result.exit_code == 0
    assert result.output == expected


@pytest.mark.parametrize(
    'entrance expected'.split(),
    [
        ({}, True),
    ],
)
def test_changelog(
    cli_runner: CliRunner,
    *,
    file_temp: Path,
    entrance: dict,
    expected: bool,
) -> None:
    """Test cli changelog."""
    entrance.update({'args': [file_temp.as_posix()]})
    assert bool(cli_runner.invoke(cli.changelog, **entrance)) == expected


incolume/py/changelog/cli.py
"""CLI - Command Line Interface module."""
from __future__ import annotations

from typing import TYPE_CHECKING

import click
from incolume.py.changelog.changelog import update_changelog

if TYPE_CHECKING:
    from pathlib import Path  # pragma: no cover


@click.command()
@click.argument('nome', envvar='USERNAME', type=click.STRING)
def greeting(nome: str) -> None:
    """Retorna a saudação para o nome passado.

    Args:
      nome: Nome de usuário

    Returns:
      Não há retorno. Uma saudação é exibida na tela.

    Raises:
      None

    Examples:
        >>> greeting Yoda
        Oi Yoda!

        >>> greeting
        Oi <usuário logado>

    """
    click.echo(f'Oi {nome.title()}!')


@click.command()
@click.argument(
    'file_changelog',
    type=click.STRING,
    default='CHANGELOG.md',
)
@click.option(
    '--url',
    '-u',
    default='https://github.com/development-incolume/'
            'incolume.py.changelog/-/compare',
    help='Url compare from repository of project.',
)
@click.option(
    '--reverse',
    '-r',
    default=True,
    help='Reverse order of records.',
)
def changelog(
    file_changelog: str | Path, url: str = '', *, reverse: bool = True,
) -> bool:
    """Operacionaliza uma interface CLI para módulo incolume.py.changelog.

    Args:
        file_changelog:  changelog full filename.
        url: url compare from repository of project.
        reverse: Reverse order of records.

    Returns:
        True if success

    Raises:
        ValueError: When there is not git tag records.

    """
    return update_changelog(
        changelog_file=file_changelog,
        urlcompare=url,
        reverse=reverse,
    )


incolume/py/changelog/__init__.py
"""Principal module."""
from __future__ import annotations

import logging
import re
from pathlib import Path
from typing import TYPE_CHECKING, Any

import toml

if TYPE_CHECKING:
    from collections.abc import Collection



confproject = Path(__file__).parents[3] / 'pyproject.toml'
versionfile = Path(__file__).parent / 'version.txt'
versionfile.write_text(
    toml.load(confproject)['tool']['poetry']['version'] + '\n',
)

__version__ = versionfile.read_text().strip()
__title__ = 'incolume.py.changelog'


def key_versions_2_sort(
    x: Collection[str], qdig: int = 0, regex: str = '',
) -> str:
    """Sort by SemVer notation.

    :param regex: regex to version format.
    :param qdig: Quantity digits to sort.
    :param x: x[key, value] -> 'git tag -ln' output
    :return: list sorted
    """
    qdig = qdig or 5
    if not isinstance(x, tuple | list):
        msg = "'x' must be tuple or list."
        raise TypeError(msg)

    classifies = {
        'post': 9 * 10**qdig,
        'rc': 8 * 10 ** (qdig - 1),
        'alpha': 2 * 10 ** (qdig - 1),
        'a': 2 * 10 ** (qdig - 1),
        'dev': 0,
    }
    regex = regex or r'(\d+)\.(\d+)\.(\d+)((-?\D+)(\d+))?'
    get_major_minor_patch_build = re.compile(regex)
    logging.debug(get_major_minor_patch_build)
    try:
        # pegar major, minor e patch
        values = get_major_minor_patch_build.search(x[0])
        major = values.group(1)
        minor = values.group(2)
        patch = values.group(3)
        build = values.group(6)
        # pegar build, se não tiver colocar uma alta 99999
        build = build or '9' * qdig
        logging.debug('values.group(5): %s', values.group(5))
        plus = classifies.get(
            re.sub(r'[-.]', '', str(values.group(5)).lower()),
            0,
        )
        logging.debug('plus: %s', plus)
        build = int(build) + plus
        result = f'{major:0>4}{minor:0>2}{patch:0>2}.{build:0>6}'
    except AttributeError:
        result = str(x[0])
    return result


def logger(str_format='', datefmt='', level=0, filelog=None):
    """Logger function for log.

    :str_format:
    :datefmt:
    :level: can be (logging.DEBUG, logging.INFO, logging.WARNING,
       logging.ERROR, logging.CRITICAL)
    :filelog:
    """
    str_format = (
        str_format
        or '%(asctime)s;%(levelname)-8s;%(name)s;'
        '%(module)s;%(funcName)s;%(message)s'
    )
    datefmt = datefmt or '%Y/%m/%d %H:%M:%S %z'
    # create logger
    level = level or logging.DEBUG
    filelog = filelog or Path(__file__).with_suffix('.py')

    logging.basicConfig(
        filename=filelog, level=level, format=str_format, datefmt=datefmt,
    )

    console = logging.StreamHandler()
    formatter = logging.Formatter(str_format)
    console.setFormatter(formatter)
    logging.getLogger('').addHandler(console)

    return logging.getLogger()


def namespace(package_name: str) -> list[str]:
    """Return the namespace.

    Example:
    package_name='incolumepy.package.module'
    ['incolumepy','incolumepy.package'].

    :param package_name: str
    :return: list

    >>> namespace('incolumepy.package.subpackage.module')
    ['incolumepy', 'incolumepy.package', 'incolumepy.package.subpackage']

    >>> namespace('incolumepy.package.module')
    ['incolumepy', 'incolumepy.package']

    >>> namespace('incolumepy.package')
    ['incolumepy']

    >>> namespace('incolumepy')
    ['incolumepy']
    """
    logging.debug('package_name=%s', package_name)
    result: list[Any] = []
    temp = ''
    try:
        bits = package_name.split('.')
    except AttributeError:
        return result

    if len(bits) <= 1:
        logging.debug('bits=%s', bits)
        return bits

    for bit in bits[:-1]:
        temp = f'{temp}.{bit}' if temp else bit
        logging.debug('temp=%s', temp)
        result.append(temp)
    logging.debug('result=%s', result)
    return result


incolume/py/changelog/changelog.py
"""Changelog Module."""
from __future__ import annotations

import inspect
import logging
import re
import subprocess
from pathlib import Path
from typing import Any

from incolume.py.changelog import __title__, __version__, key_versions_2_sort

logging.basicConfig(
    level=logging.DEBUG,
    format='%(asctime)s;%(levelname)-8s;%(name)s;'
           '%(module)s;%(funcName)s;%(message)s',
)

CHANGELOG_FILE = Path(__file__).parents[2] / 'CHANGELOG.md'


def msg_classify(msg: str, lang: str = '') -> dict[str, Any]:
    """Classify and sort one record for messages git tag -n.

    Args:
        lang: Language of command.
        msg: Message of command.
        kwargs: Anyone of the positional items.

    Returns:
        A dictionary with the version, date and a message.

    Raises:
        ValueError: If lang selected don't have support.

    Examples:

        >>> msg_classify('Adicionado: Nova funcionalidade.', 'pt-BR')
        {key: 0.1.0, date: 2023-12-18, messages: Added: Nova funcionalidade.}

        >>> msg_classify('Corregido: corrección de error.', 'es-AR')
        ValueError: es-AR not suported! Use en-US, pt-BR
    """
    logging.debug(lang)
    suport_lang: dict[str, Any] = {
        'en-US': {
            'Added': 'Added',
            'Changed': 'Changed',
            'Deprecated': 'Deprecated',
            'Removed': 'Removed',
            'Fixed': 'Fixed',
            'Security': 'Security',
        },
        'pt-BR': {
            'Adicionado': 'Added',
            'Modificado': 'Changed',
            'Obsoleto': 'Deprecated',
            'Removido': 'Removed',
            'Corrigido': 'Fixed',
            'Segurança': 'Security',
        },
    }
    suport_lang.update(
        {'all': {k: v for d in suport_lang.values() for k, v in d.items()}},
    )
    if lang not in suport_lang:
        logging.error(
            ValueError(f'{lang} not suported! Use {suport_lang.keys()}'),
        )

    key, msg = msg.split(maxsplit=1)
    ref = key + r'^{commit}'
    cmd = ['git', 'show', '-s', '--format=%cs', ref]
    logging.debug(cmd)
    date = subprocess.check_output(cmd)  # noqa: S603
    logging.debug(date)

    logging.debug('key=%s; date=%s; msg=%s', key, date, msg)
    selected_lang = suport_lang.get(lang, suport_lang['all'])
    regex: str = rf"({'|'.join(selected_lang.keys())})\s?:"

    txt = re.sub(
        regex,
        r'§§\1§:',
        msg,
        flags=re.I,
    )
    logging.debug('txt=%s', txt)
    dct: dict[str, Any] = {}
    for i, j in sorted(
        x.strip().rstrip(';').split('§:') for x in txt.strip().split('§§') if x
    ):
        dct.setdefault(selected_lang[i.capitalize()], []).extend(
            j.strip().split(';'),
        )

    return {'key': key, 'date': date, 'messages': dct}


def changelog_messages(
    *,
    text: str,
    start: int | None = None,
    end: int | None = None,
    **kwargs: str,
) -> list[tuple[str, dict[str, Any]]]:
    """Changelog messages sort and classify.

    Args:
        end: End of Message
        start: Start of Message
        text: Changelog's message

    Returns:
        return a list with a changelog menssage.

    Raises:
        None
    """
    logging.debug('parameters: (%s %s %s %s)', text, start, end, kwargs)
    lang = kwargs.get('lang', '')

    records = []
    for msg in text.strip().splitlines()[start:end]:
        logging.debug('msg=%s', msg)
        record = msg_classify(msg=msg, lang=lang)
        logging.debug('record=%s', record)
        records.append((record['key'], record))
    logging.debug('type return %s=%s', inspect.stack()[0][3], type(records))
    logging.debug('return %s=%s', inspect.stack()[0][3], records)
    return records


def changelog_header(
    url_keepachangelog: str = '',
    url_semver: str = '',
    url_convetional_commit: str = '',
) -> list[str]:
    r"""Header of changelog file.

    Args:
        url_convetional_commit: Url of convetional commit.
        url_keepachangelog: Url of keep a changelog.
        url_semver: Url of semantic version.

    Returns:
        Return a list with a header of changelog file.

    Raises:
        None

    Examples:
        >>> changelog_header()
        ['# CHANGELOG\n\n\n',
        'All notable changes to this project',
        ' will be documented in this file.\n\n',
        'The format is based on ',
        '[Keep a Changelog](https://keepachangelog.com/en/1.0.0/), ',
        'this project adheres to'
        '[Semantic Versioning](https://semver.org/spec/v2.0.0.html) '
        'and [Conventional Commit]'
        '(https://www.conventionalcommits.org/pt-br/v1.0.0/).\n\n',
        'This file was automatically generated for',
        ' [incolume.py.changelog](https://gitlab.com/development-incolume/'
        'incolumepy.utils/-/tree/0.2.0a2)',
        '\n\n---\n']

        >>> changelog_header(
            url_keepachangelog='https://keepachangelog.com/en/2.0.0/',
            url_semver=https://semver.org/spec/v1.0.0.html,
            )
        ["# CHANGELOG\n\n\n",
        "All notable changes to this project",
        " will be documented in this file.\n\n",
        "The format is based on ",
        "[Keep a Changelog](https://keepachangelog.com/en/2.0.0/),",
        "this project adheres to "
        "[Semantic Versioning](https://semver.org/spec/v1.0.0.html)"
        "and [Conventional Commit]"
        "(https://www.conventionalcommits.org/pt-br/v1.0.0/).\n\n",
        "This file was automatically generated for",
        "[incolume.py.changelog](https://gitlab.com/development-incolume/"
        "incolumepy.utils/-/tree/0.2.0a0)",
        "\n\n---\n"]
    """
    url_keepachangelog = (
        url_keepachangelog or 'https://keepachangelog.com/en/1.0.0/'
    )
    url_semver = url_semver or 'https://semver.org/spec/v2.0.0.html'
    url_convetional_commit = (
        url_convetional_commit
        or 'https://www.conventionalcommits.org/pt-br/v1.0.0/'
    )
    return [
        '# CHANGELOG\n\n\n',
        'All notable changes to this project',
        ' will be documented in this file.\n\n',
        'The format is based on ',
        f'[Keep a Changelog]({url_keepachangelog}), ',
        'this project adheres to '
        f'[Semantic Versioning]({url_semver}) '
        f'and [Conventional Commit]({url_convetional_commit}).\n\n',
        'This file was automatically generated for',
        f' [{__title__}](https://gitlab.com/development-incolume/'
        f'incolumepy.utils/-/tree/{__version__})',
        '\n\n---\n',
    ]


def changelog_body(
    content: list[tuple[str, dict[str, Any]]],
    content_formated: list[str],
    **kwargs: str,
) -> list[str]:
    """Body of changelog file.

    Args:
        content: Content of changelog.
        content_formated: Content formated of changelog.

    Returns:
        Return a list with a content of changelog's body.

    Raises:
        None
    """
    logging.debug(kwargs)
    content_formated.extend(Changelog.iter_logs(content[:-1]))
    content_formated.extend(Changelog.iter_logs(content[-1:], linked=False))
    return content_formated


def changelog_footer(
    content: list[tuple[str, dict[str, Any]]],
    content_formated: list[str],
    **kwargs: str,
) -> list[str]:
    """Footer of changelog file.

    Args:
        content: Content of changelog's footer 
        content_formated: Content formated of changelog's footer
        urlcompare: Url to compare.

    Returns:
        Return a list with a footer of changelog file.

    Raises:
        None
    """
    urlcompare = (
        kwargs.get('urlcompare')
        or 'https://gitlab.com/development-incolume/incolumepy.utils/-/compare'
    )
    logging.debug('urlcompare=%s', urlcompare)
    content_formated.append('\n---\n\n')
    y: dict[str, Any] = {}
    for _, x in content[::-1]:
        if y:
            content_formated.append(
                f'[{x["key"]}]: {urlcompare}/{y["key"]}...{x["key"]}\n',
            )
        y = x
    return content_formated


def changelog_write(
    *,
    content: list[tuple[str, dict[str, Any]]],
    **kwargs: str,
) -> bool:
    """Write CHANGELOG.md file formatted.

    Args:
        changelog_file: Path of changelog file.
        content: Content to write the changelog

    Returns:
        True if success.
        
    Raises:
        None

    Examples:

        >>> changelog_write(['Added: funcionalidade nova.'])
        True
    """
    changelog_file = Path(kwargs.get('changelog_file') or CHANGELOG_FILE)
    logging.debug('changelog_file=%s', changelog_file)

    content_formated = changelog_header()
    content_formated = changelog_body(content, content_formated, **kwargs)
    content_formated = changelog_footer(content, content_formated, **kwargs)

    with changelog_file.open('w') as f:
        f.writelines(content_formated)
        return True


def update_changelog(
    changelog_file: str = '',
    *,
    reverse: bool = True,
    **kwargs: str,
) -> bool:
    """Update Changelog.md file.

    Args:
        changelog_file: changelog full filename.
        reverse: reverse to the last update be the first.
        urlcompare(str): Url to compare. 
        
    Returns:
        True if success

    Raises:
        None

    Examples:
        >>> update_changelog()
        True
        >>> update_changelog(changelog_file='/tmp/CHANGELOG.md')
        True
        >>> update_changelog(changelog_file=Path('CHANGELOG.md'),
        urlcompare="https://gitlab.com/development-incolume
        /incolumepy.utils/-/compare")
        True
    """
    logging.debug('argumentos=%s,%s,%s', changelog_file, reverse, kwargs)
    urlcompare: str = (
        kwargs.get('urlcompare')
        or 'https://gitlab.com/development-incolume/incolumepy.utils/-/compare'
    )
    content: str = kwargs.get('content', subprocess.getoutput('git tag -n'))
    logging.info('registros encontrados ..')
    logging.debug('content=%s', content)

    return changelog_write(
        content=sorted(
            changelog_messages(
                text=content,
                start=kwargs.get('start', None),
                end=kwargs.get('end', None),
            ),
            reverse=reverse,
            key=key_versions_2_sort,
        ),
        urlcompare=urlcompare,
        changelog_file=changelog_file,
    )


class Changelog:
    """Changelog class."""

    def __init__(
        self: Changelog,
        file_output: Path | str = '',
        url_compare: str = '',
        *,
        reverse: bool = True,
        **kwargs: str,
    ):
        """Initialize from Changelog class.

        Args:
            file_output: The output file of changelog.
            reverse: reverse to the last update be the first.
            url_compare: Url to compare. 
            url_convetional_commit(str): Url of convetional commit.
            url_keepachangelog(str): Url of keep a changelog.
            url_principal(str): Url principal do projeto.
            url_semver(str): Url of semantic version.
            
        Returns:
            None
        """
        self.file_output = file_output or Path('CHANGELOG.md')
        self.url_compare = url_compare
        self.reverse = reverse
        self.url_principal = kwargs.get(
            'url_pricipal',
            'https://gitlab.com/development-incolume/incolume.py.changelog',
        )
        self.url_keepachangelog = kwargs.get(
            'url_keepachangelog', 'https://keepachangelog.com/en/1.0.0/',
        )
        self.url_semver = kwargs.get(
            'url_semver', 'https://semver.org/spec/v2.0.0.html',
        )
        self.url_convetional_commit = kwargs.get(
            'url_convetional_commit',
            'https://www.conventionalcommits.org/pt-br/v1.0.0/',
        )

    @staticmethod
    def iter_logs(
        content: list[tuple[str, dict[str, Any]]],
        *,
        linked: bool = True,
    ) -> list[str]:
        """Iterador de registros git.

        Args:
            content: Content register in git.
            linked: If has link.

        Return:
            A list with the version, date and a message.

        Raises:
            None
        """
        result = []
        for _, entrada in content:
            logging.debug(entrada)
            if linked:
                result.append(
                    f"\n\n## [{entrada['key']}]\t &#8212; "
                    f"\t{entrada['date']}:",
                )
            else:
                result.append(
                    f"\n\n## {entrada['key']}\t &#8212; \t{entrada['date']}:",
                )

            for label, msgs in entrada['messages'].items():
                result.append(f'\n### {label.capitalize()}')
                for msg in msgs:
                    frase = msg.strip()
                    frase = frase[0].upper() + frase[1:]
                    result.append(f'\n  - {frase};')
        return result

    def header(self: Changelog) -> list[str]:
        """Header of changelog file.

        Returns:
            Return a list with a header of changelog file.

        Raises:
            None

        Examples:

            >>> changelog_header()
            ['# CHANGELOG\n\n\n',
            'All notable changes to this project',
            ' will be documented in this file.\n\n',
            'The format is based on ',
            '[Keep a Changelog](https://keepachangelog.com/en/1.0.0/), ',
            'this project adheres to'
            '[Semantic Versioning](https://semver.org/spec/v2.0.0.html) '
            'and [Conventional Commit]'
            '(https://www.conventionalcommits.org/pt-br/v1.0.0/).\n\n',
            'This file was automatically generated for',
            ' [incolume.py.changelog](https://gitlab.com/development-incolume/'
            'incolumepy.utils/-/tree/0.2.0a2)',
            '\n\n---\n']

            >>> changelog_header(
                url_keepachangelog='https://keepachangelog.com/en/2.0.0/',
                url_semver=https://semver.org/spec/v1.0.0.html,
                )
            ["# CHANGELOG\n\n\n",
            "All notable changes to this project",
            " will be documented in this file.\n\n",
            "The format is based on ",
            "[Keep a Changelog](https://keepachangelog.com/en/2.0.0/),",
            "this project adheres to "
            "[Semantic Versioning](https://semver.org/spec/v1.0.0.html)"
            "and [Conventional Commit]"
            "(https://www.conventionalcommits.org/pt-br/v1.0.0/).\n\n",
            "This file was automatically generated for",
            "[incolume.py.changelog](https://gitlab.com/development-incolume/"
            "incolumepy.utils/-/tree/0.2.0a0)",
            "\n\n---\n"]
            """
        return [
            '# CHANGELOG\n\n\n',
            'All notable changes to this project',
            ' will be documented in this file.\n\n',
            'The format is based on ',
            f'[Keep a Changelog]({self.url_keepachangelog}), ',
            'this project adheres to '
            f'[Semantic Versioning]({self.url_semver}) '
            f'and [Conventional Commit]({self.url_convetional_commit}).\n\n',
            'This file was automatically generated for',
            f' [{__title__}]({self.url_principal}/-/tree/{__version__})',
            '\n\n---\n',
        ]


def run() -> None:
    """Examples ran.

    Returns:
        None
    """
    msg = subprocess.getoutput('git tag -n').splitlines()[-14]
    logging.debug(msg)
    logging.debug('msg_classify=%s', msg_classify(msg=msg))

    msg = subprocess.getoutput('git tag -n')
    result = changelog_messages(text=msg)

    logging.debug('result=%s', result)
    logging.debug('type(result)=%s', type(result))
    result = sorted(result, reverse=True, key=key_versions_2_sort)
    logging.debug('result = %s; result type = %s', result, type(result))

    changelog_write(content=result)
    update_changelog()
    logging.info(msg_classify(
        '0.1.0           added: Projeto emancipado de '
        'https://gitlab.com/development-incolume/incolumepy.utils'),
    )


if __name__ == '__main__':  # pragma: no cover
    run()
